bsip-0044.md

```diff
@@ -0,0 +1,173 @@
+
+If a satisfactory `preimage` is not provided to the HTLC before the stipulated `time lock` expires, the `depositor` may request the return of `escrow assets`. The HTLC will only evaluate transfer request from `depositor` and after `timeout threshold`, then return `escrow assets` to `depositor`.
+
+### **Condition Evaluators**
+
+The `preimage` can be thought of a secret key, that will eventually be shared with the `recipient`. This can be a word, a phrase, or even a random series of bytes. The `length` of the `preimage` must be specified within the HTLC at creation.
+
+Upon presentation of a `preimage`, the HTLC `condition evaluator` validates:
+
+1. That the `timeout threshold` has not yet occurred.
+
+2. That the length of the `preimage` matches the specified `preimage length`. 
+
+3. That the `hash of the preimage` matches the specified `preimage hash`. 
+
+If all evaluations succeed, the `escrow assets` are transferred to the `recipient`. If any evaluation fails, nothing happens; the HTLC remains ready to evaluate the next `preimage`.
+
+Upon presentation of a request from the `depositor` for return of `escrow assets`, the HTLC `condition evaluator` validates that the `timeout threshold` is now in the past. It then returns the `escrow assts` to `depositor`.
+
+### **Timing of Condition Evaluation**
+
+The `timeout threshold` of the contract is defined by `depositor` within the HTLC at creation. It can be any time in the future and should allow enough time for `recipient` to review the HTLC and provide the `preimage`. Further, it should not be set too far into the future to mitigate against an unresponsive `recipient` impacting `depositor`, as their `escrow assets` will be locked until `timeout threshold` expiry. The accuracy is based on when the `condition evaluator` runs, and should be considered accurate ± 15 seconds.
+
+### **Early Termination of an Escrow Contract**
+
+To protect the `recipient`, early termination of an HTLC is not allowed by any party. Placing a `timeout threshold` far into the future is valid. User protection from locking up funds for an extremely long period could be provided by the UI used to create the HTLC.
+
+### **Automatic Transfers Upon Expiry**
+
+Upon expiry of the `timeout threshold`, the `escrow assets` held within the HTLC will **not** automatically be returned to `depositor`. A separate transaction must be initiated by the `depositor` to request `escrow assets` be transferred to `depositor`. 
+
+Code _could_ be added to automate the return of funds. This could be part of block production or maintenance. However, this will cause extra load on those processes. Hence, it is felt that requiring action by the `depositor` is the best option.
+
+### **Fee**
+
+Creating and fulfillment are two operations that add data to the blockchain. The `fee` for each operation is based on the standards set for blocks, and is similar to costs of other items stored on-chain.
```

Comment:
Please be more specific. There's a difference between creating and fulfilling.
* Creating consumes a more or less fixed amount of memory for a significant time period. The is comparable to creating limit orders (but requires more memory, depending on hash length). It should therefore come with a significant fixed fee greater than that of a limit order.
* Fulfilling frees up memory, but a long pre-image consumes space in a block and network bandwidth. It should therefore come with a small fixed fee, plus a per-kb fee.

Reply:
Consider the following as a replacement:
### **Fees**

We propose two (2) operations (see Specification) to implement the HTLC feature, each requiring distinct fees. All fees will be set and maintained by the Committee. 

The "prepare" operation will store in-memory data on validation nodes until redeemed or expires. We recommend this operation have two (2) fees: `prepare_fee` which is flat and `prepare_durration_day_fee` which is calculated by based on the number of days until `timeout threshold`. 

The "redeem" operation frees the memory from the validation nodes and adds the `preimage` to blockchain data storage. We recommend this operation have two (2) fees: `redeem_fee` which is quite low and `redeem_preimage_kb_fee` which is based the amount of data committed to the blockchain.

Reply:
Also intend to add notes prior to the fee section:

**Note:** we recommend the Committee the set maximum allowable `preimage length` to ensure unreasonably large submissions are rejected. 

**Note:** we recommend the Committee set the maximum value for `timeout threshold` to limit the amount of time a contract may consume memory of validation nodes.

Reply:
Please clarify: If an automatic return of funds is implemented, when it is used, no fee is created. Correct? Having a fee on an "automatic" operation gets messy.

Reply:
Correct, automatic return of funds should be free of charge.
Fee per day should work too.